File: ext/post-to-patreon.js
import fs from 'fs';
import axios from 'axios';

const JSON_FILE = 'artikel.json';
const DATABASE_FILE = 'mini/posted-patreon.txt';
const BASE_URL = 'https://dalam.web.id';

const slugify = (text) =>
  text.toLowerCase().trim().replace(/\s+/g, '-');

async function postToPatreon() {
  const ACCESS_TOKEN = process.env.PATREON_ACCESS_TOKEN;
  const CAMPAIGN_ID = process.env.PATREON_CAMPAIGN_ID || "15483016"; // Pakai ID yang sudah kita temukan tadi

  if (!ACCESS_TOKEN) {
    console.error('❌ Patreon token belum diset');
    process.exit(1);
  }

  if (!fs.existsSync(JSON_FILE)) {
    console.error('❌ File artikel.json tidak ditemukan');
    return;
  }

  const data = JSON.parse(fs.readFileSync(JSON_FILE, 'utf8'));
  let postedDatabase = fs.existsSync(DATABASE_FILE) ? fs.readFileSync(DATABASE_FILE, 'utf8') : "";
  let allPosts = [];

  for (const [cat, posts] of Object.entries(data)) {
    const catSlug = slugify(cat);
    posts.forEach(p => {
      const fileSlug = p[1].replace('.html', '').replace(/^\//, '');
      const fullUrl = `${BASE_URL}/${catSlug}/${fileSlug}/`;

      if (!postedDatabase.includes(fileSlug)) {
        allPosts.push({
          title: p[0],
          slug: fileSlug, // Kita simpan slug saja di db biar rapi
          url: fullUrl,
          date: p[3],
          desc: p[4] || "Artikel terbaru dari dalam.web.id"
        });
      }
    });
  }

  allPosts.sort((a, b) => b.date.localeCompare(a.date));

  if (allPosts.length === 0) {
    console.log("🏁 Patreon: Semua artikel sudah dipost.");
    return;
  }

  const target = allPosts[0];
  console.log(`🚀 Posting ke Patreon: ${target.title}`);

  try {
    const response = await axios.post(
      // KITA GUNAKAN RUTE SPESIFIK KAMPANYE
      `https://www.patreon.com/api/oauth2/v2/campaigns/${CAMPAIGN_ID}/posts`,
      {
        data: {
          type: 'post',
          attributes: {
            title: target.title,
            content: `<p>${target.desc}</p><p>🔗 Kupas Tuntas di: <a href="${target.url}">${target.url}</a></p>`,
            is_public: true,
            publish_state: 'published'
          }
          // Di rute ini, kita TIDAK PERLU lagi mengirim relationships campaign
          // karena ID kampanye sudah ada di URL.
        }
      },
      {
        headers: {
          'Authorization': `Bearer ${ACCESS_TOKEN}`,
          'Content-Type': 'application/vnd.api+json',
          'Accept': 'application/vnd.api+json'
        }
      }
    );

    if (!fs.existsSync('mini')) fs.mkdirSync('mini', { recursive: true });
    fs.appendFileSync(DATABASE_FILE, target.slug + '\n');

    console.log(`✅ AKHIRNYA BERHASIL! "${target.title}" sudah tayang.`);
  } catch (err) {
    // JIKA RUTE KAMPANYE MASIH 404, KITA COBA FALLBACK KE RUTE UMUM (UNTUK BERJAGA-JAGA)
    if (err.response?.status === 404) {
        console.log("🔄 Rute kampanye 404, mencoba rute umum v2 sebagai cadangan...");
        // ... (Logika fallback jika diperlukan)
    }

    console.error(
      '❌ Patreon Error:',
      JSON.stringify(err.response?.data || err.message, null, 2)
    );
    process.exit(1);
  }
}

postToPatreon();

File: ext/post-to-patreon.py
import json
import os
import sys
import re
import requests

# Konfigurasi Path & URL
JSON_FILE = 'artikel.json'
DATABASE_FILE = 'mini/posted-patreon.txt'
DOMAIN_URL = 'https://dalam.web.id'

# Patreon Config (Ambil dari GitHub Secrets)
PATREON_ACCESS_TOKEN = os.getenv('PATREON_ACCESS_TOKEN')
CAMPAIGN_ID = os.getenv('PATREON_CAMPAIGN_ID')

def slugify(text):
    text = text.strip().lower()
    text = re.sub(r'\s+', '-', text)
    return text

def main():
    if not os.path.exists(JSON_FILE):
        print("Error: artikel.json tidak ditemukan")
        sys.exit(1)

    # Load data artikel
    with open(JSON_FILE, 'r', encoding='utf-8') as f:
        data = json.load(f)

    # Load database biar nggak posting ulang
    posted_database = ""
    if os.path.exists(DATABASE_FILE):
        with open(DATABASE_FILE, 'r', encoding='utf-8') as f:
            posted_database = f.read()

    all_posts = []
    for category_name, posts in data.items():
        cat_slug = slugify(category_name)
        for post in posts:
            file_name = post[1].strip()
            file_slug = file_name.replace('.html', '').replace('/', '')
            full_url = f"{DOMAIN_URL}/{cat_slug}/{file_slug}/"

            if file_slug not in posted_database:
                all_posts.append({
                    'title': post[0],
                    'slug': file_slug,
                    'url': full_url,
                    'date': post[3],
                    'desc': post[4] or "Kupas Tuntas di Layar Kosong.",
                    'category': category_name
                })

    # Urutkan dari yang terbaru
    all_posts.sort(key=lambda x: x['date'], reverse=True)

    if all_posts:
        target_post = all_posts[0]
        title = target_post['title']
        body = f"<p>{target_post['desc']}</p><p>Kupas Tuntas di: <a href='{target_post['url']}'>{target_post['url']}</a></p>"

        # Token & ID Cleaning
        token = str(PATREON_ACCESS_TOKEN).strip() if PATREON_ACCESS_TOKEN else None
        camp_id = str(CAMPAIGN_ID).strip() if CAMPAIGN_ID else None

        if not token or not camp_id:
            print("⚠️ Error: PATREON_ACCESS_TOKEN atau PATREON_CAMPAIGN_ID tidak ditemukan!")
            return

        # --- PREPARE PAYLOAD ---
        payload = {
            "data": {
                "type": "post",
                "attributes": {
                    "title": title,
                    "content": body,
                    "is_paid": False,
                    "is_public": True
                },
                "relationships": {
                    "campaign": {
                        "data": {
                            "type": "campaign",
                            "id": camp_id
                        }
                    }
                }
            }
        }

        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/vnd.api+json",
            "Accept": "application/vnd.api+json"
        }

        # --- TRY POSTING ---
        print(f"🚀 Mencoba rute kampanye untuk: {title}...")
        api_url = f"https://www.patreon.com/api/oauth2/v2/campaigns/{camp_id}/posts"

        try:
            response = requests.post(api_url, json=payload, headers=headers)

            # Jika rute kampanye ditolak (405/404), coba rute umum
            if response.status_code in [404, 405]:
                print("⚠️ Rute kampanye ditolak, mencoba rute /v2/posts...")
                api_url = "https://www.patreon.com/api/oauth2/v2/posts"
                response = requests.post(api_url, json=payload, headers=headers)

            if response.status_code in [200, 201]:
                print(f"✅ Berhasil posting ke Patreon: {title}")
                with open(DATABASE_FILE, 'a', encoding='utf-8') as f:
                    f.write(target_post['slug'] + '\n')
            else:
                print(f"❌ Gagal total. Status: {response.status_code}")
                print(f"Detail: {response.text}")

        except Exception as e:
            print(f"❌ Terjadi kesalahan saat request: {str(e)}")

    else:
        print("✅ Tidak ada artikel baru untuk diposting.")

if __name__ == "__main__":
    main()
